### ISSUE TITLE
Different results on subsets / using parallel option

### ISSUE THREAD

**Issue by GM-Alex** [squizlabs/PHP_CodeSniffer#2179]
If I run phpcs over the whole project folder (src) it doesn't find any issues but if I run it over a subset (src/subfolder) it will find the expected issues. It's the same if I use the parallel option, if I use it the expected issues will be found if not everything seems fine.

**Comment by GM-Alex** [comment]
Ok I did some research and it seems that the rule set isn't independent (I'm currently not sure if that is a ruleset problem or a general phpcs problem). If I change the File.php constructor to the following I always get the expected results:

```
/**
     * Constructs a file.
     *
     * @param string                   $path    The absolute path to the file to process.
     * @param \PHP_CodeSniffer\Ruleset $ruleset The ruleset used for the run.
     * @param \PHP_CodeSniffer\Config  $config  The config data for the run.
     *
     * @return void
     */
    public function __construct($path, Ruleset $ruleset, Config $config)
    {
        $this->path    = $path;
        $this->ruleset = new Ruleset($config);
        $this->config  = $config;
        $this->fixer   = new Fixer();
        ...
```
So as you can see I throw the existing ruleset away and regenerate a new one from the config. The problem should be that the ruleset is a reference and shared between the files.

**Comment by jrfnl** [comment]
Are you using a custom ruleset and if so, could you please post the ruleset you are using ?

**Comment by GM-Alex** [comment]
Yes I'm using https://github.com/bestit/PHP_CodeSniffer with in a spryker project which comes with it's own ruleset, see https://github.com/spryker/code-sniffer. The `phpcs.xml.dist` looks like this:

```
<?xml version="1.0"?>
<ruleset name="MyRuleset">
    <file>src</file>

    <arg name="basepath" value="."/>
    <arg name="colors" />
    <!-- Suppress warnings, show sniff source and progress -->
    <arg value="nsp"/>

    <!-- Lint PHP files -->
    <rule ref="Generic.PHP.Syntax"/>

    <rule ref="./vendor/spryker/code-sniffer/Spryker/ruleset.xml">
        <!-- Conflict with best it rules -->
        <exclude name="Spryker.Commenting.DocBlockTagOrder" />
        <exclude name="Spryker.Commenting.FileDocBlock.CustomFileDocBlockMissing" />
        <exclude name="Spryker.Commenting.FullyQualifiedClassNameInDocBlock" />
        <exclude name="Spryker.WhiteSpace.OperatorSpacing" />
        <exclude name="Generic.Formatting.NoSpaceAfterCast" />
        <exclude name="SlevomatCodingStandard.Namespaces.UnusedUses"/>
        <exclude name="Spryker.Commenting.DocBlockReturnVoid.ReturnMissingInInterface"/>
        <exclude name="Spryker.Commenting.DocBlockNoInlineAlignment.WhitespaceAroundTypes"/>
        <exclude name="SlevomatCodingStandard.Namespaces.UseFromSameNamespace"/>
        <exclude name="SlevomatCodingStandard.Namespaces.AlphabeticallySortedUses"/>
    </rule>

    <rule ref="./vendor/bestit/php_codesniffer/src/Standards/BestIt/ruleset.xml">
        <!-- Contains a bug which leads to PHP syntax errors, add it back once that's fixed (PHPCS-73) -->
        <exclude name="BestIt.TypeHints.ReturnTypeDeclaration"/>
        <!-- Currently has a bug and doesn't think "SomeObject[]" is valid. -->
        <exclude name="BestIt.DocTags.VarTag.TagContentFormatInvalid"/>
        <!-- This sniff does not account for usages in DocBlocks -->
        <exclude name="SlevomatCodingStandard.Namespaces.UnusedUses"/>
    </rule>

    <rule ref="BestIt.DocTags.DisallowedClassTags">
        <properties>
            <property name="disallowedTags" type="array">
                <element value="category"/>
                <element value="copyright"/>
                <element value="example"/>
                <element value="filesource"/>
                <element value="global"/>
                <element value="ignore"/>
                <element value="internal"/>
                <element value="license"/>
                <element value="param"/>
                <element value="property"/>
                <element value="property-read"/>
                <element value="property-write"/>
                <element value="return"/>
                <element value="since"/>
                <element value="source"/>
                <element value="subpackage"/>
                <element value="throws"/>
                <element value="todo"/>
                <element value="uses"/>
                <element value="var"/>
                <element value="inheritdoc"/>
            </property>
        </properties>
    </rule>

    <rule ref="BestIt.DocTags.DisallowedMethodTags">
        <properties>
            <property name="disallowedTags" type="array">
                <element value="author"/>
                <element value="category"/>
                <element value="copyright"/>
                <element value="filesource"/>
                <element value="global"/>
                <element value="ignore"/>
                <element value="internal"/>
                <element value="license"/>
                <element value="link"/>
                <element value="method"/>
                <element value="package"/>
                <element value="property"/>
                <element value="property-read"/>
                <element value="property-write"/>
                <element value="since"/>
                <element value="source"/>
                <element value="subpackage"/>
                <element value="todo"/>
                <element value="uses"/>
                <element value="var"/>
                <element value="version"/>
                <element value="inheritdoc"/>
            </property>
        </properties>
    </rule>

    <exclude-pattern>*/src/Generated/*</exclude-pattern>
    <exclude-pattern>*/src/Orm/*/Base/</exclude-pattern>
    <exclude-pattern>*/src/Orm/*/Map/</exclude-pattern>
    <exclude-pattern>*/src/Orm/Propel/</exclude-pattern>
    <exclude-pattern>*/tests/_support/_generated/*</exclude-pattern>
    <exclude-pattern>*/tests/_helpers/*</exclude-pattern>
    <exclude-pattern>*/tests/_output/*</exclude-pattern>
    <exclude-pattern>./data/DE/*</exclude-pattern>
    <exclude-pattern>./data/AT/*</exclude-pattern>
    <exclude-pattern>./data/US/*</exclude-pattern>
    <exclude-pattern>*/node_modules/*</exclude-pattern>
</ruleset>
```

**Comment by jrfnl** [comment]
@GM-Alex Got some more info on the project you are running this on ? Like what is the basepath ? and how does the directory structure look like ?

Both the above ruleset + the Spyker ruleset contain quite a few `exclude-pattern`s, so that would be the first thing I would suggest to look at.
Keep in mind that those patterns are interpreted as regular expressions (with any `*` replaced by `.*`).

**Comment by GM-Alex** [comment]
@jrfnl What I think is strange is that the ruleset isn't independent and changes its behaviour related to the files which are already processed. That shouldn't be the case anyway in my opinion, but to be honest I'm not a phpcs pro so it could be that sniffs must be "stateless" and the sniffs are storing some kind of state internals, so the result is that uncontrollable behaviour.

**Comment by jrfnl** [comment]
> so the result is that uncontrollable behaviour.

@GM-Alex Well, I've never seen that kind of behaviour which wasn't caused by botched exclude pattern, so I'm not sure what you mean.

**Comment by jrfnl** [comment]
@GM-Alex I'm not sure I understand what you mean. Is there some way I could try to reproduce the behaviour ?

**Comment by GM-Alex** [comment]
@jrfnl Currently not because it's an none public project, but the problem is the following. File `a.php` has an issue lets say a missing `@var` tag. The project structure is the following:

```
-src
--a
---a.php
---somefiles...
--b
---somefile 
```

If I run it with the config file above no error will shown, if I change `<file>src</file>` to `<file>src/a</file>` the error will shown and if I run phpcs src/a/a.php I also see the error.

**Comment by jrfnl** [comment]
@GM-Alex Just out of curiousity, not sure if it will make any difference at all, but could you try changing the `<file>` directive to `<file>./src/</file>` and if that still yields the same result, to `<file>src/</file>` ?
I'm wondering if it's about PHPCS recognizing something as a directory vs a file.

**Comment by GM-Alex** [comment]
@jrfnl I already changed that and tested several combinations, so also the the code modification I made above fixes the issues, so it has definitely something to do the the ruleset.

**Comment by gsherwood** [comment]
> The reason is that the ruleset changes related to the file which where processed during the runtime

The ruleset should not be changing during a run, and should be identical for all files. I suspect the exclude patterns are interfering with things, but that's pretty much impossible to tell unless you can provide steps to replicate.

**Comment by DaveLak** [comment]
@jrfnl @gsherwood

I am seeing a similar issue with PHPCS generating different output when run against a directory vs running it against a single file in the same directory. I've put together [a little POC repo that's reproducing the issue (on my machine at least)](https://github.com/DaveLak/phpcs-poc).

I'm not sure if it's a PHPCS issue or an issue with the [WordPress-Coding-Standards](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards) ruleset I'm using though. If you could have a look and confirm or rule out it being an issue in PHPCS itself that would be greatly appreciated.

**Comment by DaveLak** [comment]
@jrfnl Thanks for taking a look so quickly. Screenshot of my shell in the foldout.

> What OS are you on ? And which PHP version are you using ?

* OS: macOS Mojave 10.14.1
* PHP: PHP 7.1.19



<details>
  <summary>Screenshot</summary>
 
  ---

  <img width="722" alt="phpcs-poc" src="https://user-images.githubusercontent.com/17415134/48308123-a9812380-e52a-11e8-832f-2496a265479d.png">

  ---

Also, this collapsable content is neat! I didn't realize it was possible.
</details>

**Comment by jrfnl** [comment]
@DaveLak Ok, I've just tested again on my Windows 7 machine with PHP 7.1.19 and still cannot reproduce the issue. So, yes, I think this may be a Mac OS related issue. Unfortunately I don't have a mac available to test with, so I can't help you any further than this.

**Comment by DaveLak** [comment]
No worries, thanks for taking a look @jrfnl

**Comment by gsherwood** [comment]
@DaveLak @jrfnl I can replicate this on Mac OS.

The issue seems to be with the structure of the `$ignoreAlignmentTokens` member var of the `WordPress.WhiteSpace.PrecisionAlignment` sniff. When run individually, the array looks like this:
```
Array
(
    [0] => T_COMMENT
    [1] => T_INLINE_HTML
)
```
While when run as a batch, the array looks like this:
```
Array
(
    [T_COMMENT] =>
    [T_INLINE_HTML] =>
)
```
Interestingly, the first array format (where it doesn't produce errors) appears like that for `functions.php` but not for `category.php`. I haven't figured out why yet.

**Comment by gsherwood** [comment]
@DaveLak @jrfnl Seems to be just be the usage of `merge_custom_array()`.

When run together you can see that the sniff's member var starts out with the token types as values. Then it converts to token types being used as keys the first time the sniff runs. On the second (and subsequent) files, the array format is now different and `merge_custom_array()` makes it blank:
```
Processing functions.php [PHP => 138 tokens in 18 lines]...
SNIFF HAS THIS ARRAY VALUE:
Array
(
    [0] => T_COMMENT
    [1] => T_INLINE_HTML
)
MERGE PRODUCES THIS ARRAY VALUE:
Array
(
    [T_COMMENT] =>
    [T_INLINE_HTML] =>
)
DONE in 85ms (0 errors, 2 warnings)
Processing category.php [PHP => 25 tokens in 14 lines]...
SNIFF HAS THIS ARRAY VALUE:
Array
(
    [T_COMMENT] =>
    [T_INLINE_HTML] =>
)
MERGE PRODUCES THIS ARRAY VALUE:
Array
(
)
```

Changing this code:
```php
// Handle any custom ignore tokens received from a ruleset.
$this->ignoreAlignmentTokens = $this->merge_custom_array( $this->ignoreAlignmentTokens );
```
To instead use a local var fixes this. When run together, the category.php file does not produce any warnings.

**Comment by jrfnl** [comment]
@gsherwood Thank you for discovering that. Oopsidaisy, that wasn't supposed to happen... This should be fixed now through PR https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/1532

@DaveLak If you'd have a chance to test that PR, your feedback would be appreciated.

I do still think it's weird that I couldn't reproduce the issue on a Windows machine using the POC provided. Anyone has any ideas about that ?

**Comment by gsherwood** [comment]
> I do still think it's weird that I couldn't reproduce the issue on a Windows machine using the POC provided. Anyone has any ideas about that ?

Yeah that's an interesting one. I didn't see anything that would work differently on Windows

**Comment by gsherwood** [comment]
Closing due to lack of ability to replicate the original issue. If you are able to provide steps to help, please comment here and I will reopen.

**Comment by flaviovs** [comment]
I understand that this is an old, closed issue, but I am experiencing this every now and then, so here are my findings in hope they help find and tackle the bug:

* Running `phpcs .` in my local environment ends without errors. The same command when run by our CI tool (GitLab) fails for some non-compliant file.

* However, running `phpcs path/to/noncompliant.php` in my local environment warns about all CS error in the file.

* Both local environment and CI environment use the same Docker image. Host software stack is the same (Debian Buster).

* It seems to be related to parallel jobs. `phpcs --parallel=X` works as expected (finds the offending file) for X=0 and X=1 in my local 4-cores machine. Any other values for X cause offending files to be missed. Just `phpcs` seems to always work as expected in our 8-cores CI server (i.e. it always find all CS violations).

* Timing runs with time(1) shows me that `phpcs --parallel=X` for X={0, 1} takes ~12s, whereas `phpcs --parallel=2` or just `phpcs` (both miss offending files) finish in ~6s. So I assume that it is trying to go parallel by default.

* Testing with `-v` shows that files are really being ignored:

      $ ./vendor/bin/phpcs -v | grep NonCompliant

      $ ./vendor/bin/phpcs -v --parallel=1 | grep NonCompliant
      Processing NonCompliant.php [PHP => 191 tokens in 36 lines]... DONE in 3ms (1 errors, 0 warnings)
      FILE: /path/to/NonCompliant.php

* We were having this issue with v3.4.1 originally. Upgrading to latest (`PHP_CodeSniffer version 3.5.4 (stable) by Squiz (http://www.squiz.net)`) didn't solve it.

* phpcbf behaves identically. It does nothing if invoked as just `phpcbf`. It finds and fix all issues if I use `phpcbf --parallel=1`.

* This problem seems to happen randomly. Most of time both local and CI tools are "in sync" -- meaning both find no errors, or the same CS errors. But when it happens, it happens consistently until we fix the offending file and push. I couldn't determine what triggers the problem.
